Sources/FirebaseCore/FirebaseError.swift

```diff
@@ -1,12 +1,5 @@
 // SPDX-License-Identifier: BSD-3-Clause
 
-public struct FirebaseError: Error {
-  public let code: CInt
-  public let message: String
-
-  @_spi(FirebaseInternal)
-  public init(code: CInt, message: String) {
-    self.code = code
-    self.message = message
-  }
+public protocol FirebaseError: Error {
```

Comment:
I'm not sure I understand what purpose this serves. This change effectively is replacing `FirebaseError` with `AuthErrorCode`, I think that we should remove this protocol.

Reply:
This is a way to specify that `AuthErrorCode` and `FirestoreErrorCode` should provide an `init` method with these parameters.

Reply:
Why is that better than using a `CustomStringConvertible` conformance on the error type rather than passing that in the initialiser?

Reply:
I don't understand what you are suggesting. Firebase produces errors of the form `(number, string)`. There isn't a way to just get the `string` bit given only the `number` portion. So we need to package up both `number` and `string` in some `Error`-type to pass back to Swift.

Reply:
Okay, so the error messages are bespoke and not a templated message? The error code is not something that is related to the message at all? I think that deserves a comment with the protocol explaining how the error messages are completely disjoint from the code.

Reply:
Honestly, I can't tell. It is dependent on the implementation of firebase-cpp-sdk.

This is how we get error codes and error strings:

![image](https://github.com/thebrowsercompany/swift-firebase/assets/495958/257ef36d-4470-4e1d-8c91-1511fcab8b74)

Also even if there were a mapping from `number` to `string` here and we could get away with not passing the message to this error initializer, it seems like we'd still need to define an initializer that takes an `Int`, right?

I don't understand how avoiding the message would allow us to avoid the need for this protocol. Can you explain your thinking?

Reply:
I was going by the message from you:

> This is a way to specify that AuthErrorCode and FirestoreErrorCode should provide an init method with these parameters.

I'm not seeing what value this particular initializer requirement is adding. The initializer doesn't really matter too much, and I think that we can have a much better representation for the type using type specific initializers, particularly if we are simply bridging the type over. In theory, we could even just do an `Error` conformance on the raw C++ type if we really wanted to bubble the error up.

Since this is trying to mimic the Firebase iOS SDK type hierarchy, unless there is a `FirebaseError` type there that adds this requirement, we should simply remove this protocol.

Reply:
Here is the compilation error I get when I try to not have this protocol:

```
PS C:\workspace\swift-firebase> swift build
Building for debugging...
C:\workspace\swift-firebase\Sources\FirebaseCore\FutureProtocol.swift:53:20: error: type 'ErrorType' has no member 'init'
51 │     let error = error()
52 │     guard error == 0 else {
53 │       return (nil, ErrorType(code: error, message: errorMessage!))
   │                    ╰─ error: type 'ErrorType' has no member 'init'
54 │     }
55 │     return (result, nil)
[5/6] Compiling FirebaseCore FutureProtocol.swift
```

What are you suggesting as an alternative fix?

Reply:
You are getting that error message because `FirebaseError.init(code:message:)` is marked as `FirebaseInternal` (https://github.com/thebrowsercompany/swift-firebase/blob/main/Sources/FirebaseCore/FirebaseError.swift#L8). You need to ensure that the `import` is marked as a SPI import so that you can access the initializer.

Reply:
I don't think that's the reason why. But no need to resolve that as I have removed the `FirebaseError` protocol in the latest commit. Now `resultAndError` takes a closure that returns an `ErrorType` so that avoids the need for this internal protocol.